### PR TITLE
feat: add funcs related to unicode conversion

### PR DESCRIPTION
**Pull request by bhajneet**
### Summary

Adds a unicode function which will convert ASCII and normalize unicode strings of Gurmukhi. Works on ASCII, Unicode, and mixed (ASCII+Unicode) Gurmukhi strings.

Adds a plethora of tests to confirm compliance with Unicode for most cases. Edge cases of half-Y and 2x non-existing top-open Y chars have also been factored in.

Has the option to convert to a special "Sant Lipi" Unicode standard. This should fix the Yayya issues with Unicode by shifting code points to the Private Use Area. This would be the recommended way of storing data for the Shabad OS Database in conjunction with the Sant Lipi font.

More testing is required with the Sant Lipi font to see if ligature rules can handle half-Y + diacritics (though PUA still seems preferable to me).

This took a very long time for me to finish, I added comments where I believe most people unaware of Unicode Gurmukhi and it's peculiarities would benefit from learning about.

### Test

- Open `Gurbani Lipi`, `Anmol Lipi`, `Gurbani Akhar`, `Open Gurbani Akhar` fonts in Glyphs app to map every possible ASCII character/ligature.
- Use regex searches on the Shabad OS Database for complex vowel arrangements.
- Read through various Unicode documents including proposals.
- Over 100 assertions for pytest including diacritical ordering, complex diacritics (e.g. triple diacritics on ਕ੍ਰਾਂ, ਸ੍ਵਾਂਤਿ, ਭ੍ਰਿੰਗ), ASCII sihari in unicode strings (`ਮiਲ -> ਮਿਲ`), best-conversions for words Unicode can't handle (e.g. `iBÎo -> ਭਿਯੋ`, uses full yayya since Unicode can't compute).

### Duration

Easily 20 hours. Somewhere between 20-30 hours of work.

### CODE REVIEW

File: gurmukhiutils/unicode.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        '੧੨੩'
+
+        >>> unicode("<> > <")
+        'ੴ ☬ ੴ'
+
+        >>> unicode("gurU")
+        'ਗੁਰੂ'
+    """
+
+    import re
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:32]: imports within functions like this are generally discouraged without good reason
https://peps.python.org/pep-0008/#imports

File: gurmukhiutils/unicode.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+
+    # AnmolLipi/GurbaniAkhar & GurbaniLipi by Kulbir S. Thind, MD
+    common_ascii_to_unicode_map = {
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:35]: Constants are usually defined to be all caps
https://peps.python.org/pep-0008/#constants

You might want something like an encoding dir with unicode.py and constants.py and do something like
from gurmukhiutils.encoding import unicode, constants

File: gurmukhiutils/unicode.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        "\uee85": "₅",
+        "\uee86": "₆",
+        "\uee87": "₇",
+        "\uee88": "₈",
+        "\uee89": "₉",
+    }
+
+    # Move ASCII sihari before mapping to unicode
+    ascii_base_letters = "a-zA-Z\\|^&Îîï"
+    pattern = r"(i)([" + ascii_base_letters + "])"
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:199]: fyi you can do
`fr"(i)([{ascii_base_letters}])"`
you generally want to avoid string concatenation because python isn't type safe, f-strings or .format are useful things for this.

File: gurmukhiutils/unicode.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+    string = re.sub(pattern, r"\2\1", string)
+
+    # Replace any ASCII with Unicode Gurmukhi
+    for key, value in common_ascii_to_unicode_map.items():
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:203]: Might be worth creating a translation table for these:

https://docs.python.org/3/library/stdtypes.html?highlight=translate#str.translate

File: gurmukhiutils/unicode.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+        string = string.replace(key, value)
+
+    for key, value in special_sant_lipi_map.items():
+        string = string.replace(key, value)
+
+    # Normalize Unicode
+    string = unicode_normalize(string)
+
+    if unicode_standard == "Unicode Consortium":
+        import re
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:213]: duplicate import

File: gurmukhiutils/unicode.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
+from typing import Literal, Match
+
+UNICODE_STANDARDS = Literal["Unicode Consortium", "Sant Lipi"]
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:3]: an enum might make more sense here a little easier to compare and more explicit in the code

File: gurmukhiutils/unicode.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    Since otherwise, surrounds each lookup of a subjoined letter with lookups of the rest of the diacritics (which are single char).
+
+    The patterns for the single-chars and the subjoined letters:
+    """
+    generated_single_chars = "".join(
+        nukta_udaat_yakash_order_list
+        + vowel_order_list
+        + remaining_diacritic_order_list
+    )
+    d_pattern = "([" + generated_single_chars + "]*)"
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:331]: f-string here as well

File: gurmukhiutils/unicode.py [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+
+    subjoin_constructor = "੍"
+    subjoin_consonants = "ਹਰਵਟਤਨਚ"
+    s_pattern = "(" + subjoin_constructor + "[" + subjoin_consonants + "])?"
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:335]: f-string

File: gurmukhiutils/unicode.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+
+    """
+    The following regex will capture all sequential diacritics containing at most one subjoined letter.
+        >>> print(regex_match_pattern)
+        '([਼ੵਿੇੈੋੌੁੂਾੀਁੱਂੰਃ]*)(੍[ਹਰਵਟਤਨਚ])?([਼ੵਿੇੈੋੌੁੂਾੀਁੱਂੰਃ]*)'
+    """
+    regex_match_pattern = d_pattern + s_pattern + d_pattern
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:342]: f-string

File: gurmukhiutils/unicode.py [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+        + [subjoin_consonants]
+        + vowel_order_list
+        + remaining_diacritic_order_list
+    )
+
+    def regex_sort_func(match: Match) -> str:
+        """
+        This re-arranges characters in "match" according to a custom sort order "generated_diacritic_order"
+        """
+        if len(match := match.group()) > 1:
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:361]: you generally want to avoid using the same variable name in a function as a parameter. renaming the local match to _match is usually a quick and easy fix.

File: gurmukhiutils/unicode.py [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+    Examples:
+        >>> encode_unicode(["0054"])
+        'T'
+
+        >>> encode_unicode(["0a1c", "0A40"])
+        '['ਜ', 'ੀ']'
+    """
+    encoded_list = []
+    for string in strings:
+        encoded_list.append(chr(int(string, 16)))
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:446]: you can use a list comprehension here
`return [chr(int(string, 16)) for string in strings]`

File: gurmukhiutils/unicode.py [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+
+    Example:
+        >>> decode_unicode("To ਜੀ")
+        [{'T': '0054'}, {'o': '006f'}, {' ': '0020'}, {'ਜ': '0a1c'}, {'ੀ': '0a40'}]
+    """
+    list = []
+    for item in string:
+        list.append({item: format(ord(item), "04x")})
+
+    return list
```
Comment (FunkmasterBadass) [inline gurmukhiutils/unicode.py:424]: `return [{item: format(ord(item), "04x")} for item in string]`

**Review by FunkmasterBadass — COMMENTED** [review]
tests look good too, I would look at adding some negative cases, where you pass in values that shouldn't be converted and verify they were left alone

File: gurmukhiutils/unicode.py (hunk 4 above)
Comment (bhajneet) [reply gurmukhiutils/unicode.py:203]: I believe that only works with single characters. I need to be able to translate substrings to single/many chars. Thoughts?

File: gurmukhiutils/unicode.py (hunk 10 above)
Comment (bhajneet) [reply gurmukhiutils/unicode.py:361]: 👍 Okay, I will do that. What's the underscore indicate?
